File: frontend/src/schemas/inventory.ts
/**
 * Zod schemas — Clinic Inventory (mirrors backend InventoryController rules).
 */
import { z } from 'zod';

export const inventoryItemLastMovementSchema = z.object({
  reason_code: z.string(),
  qty_delta: z.number().int(),
  created_at: z.string(),
  user_email: z.string().nullable(),
});
export type InventoryItemLastMovement = z.infer<typeof inventoryItemLastMovementSchema>;

export const inventoryItemSchema = z.object({
  id: z.number().int().positive(),
  sku: z.string(),
  name: z.string(),
  unit: z.string(),
  quantity_on_hand: z.number().int().min(0),
  reorder_level: z.number().int().min(0),
  low_stock: z.boolean(),
  archived: z.boolean().optional().default(false),
  created_at: z.string(),
  // Row-level "last move" hint — null when the item has no movements yet.
  last_movement: inventoryItemLastMovementSchema.nullable(),
});
export type InventoryItem = z.infer<typeof inventoryItemSchema>;

export const createItemSchema = z.object({
  sku: z.string().min(1).max(64).regex(/^[A-Za-z0-9_-]+$/, 'Letters, digits, - and _ only'),
  name: z.string().min(1).max(128),
  unit: z.string().min(1).max(32).default('pc'),
  reorder_level: z.number().int().min(0).default(0),
});
export type CreateItemInput = z.infer<typeof createItemSchema>;

/**
 * Update a supply item. SKU is NOT editable (it backs the
 * movement ledger); name, unit, reorder_level are mutable. Stock
 * is untouched here — use the Move dialog for that.
 */
export const updateItemSchema = z.object({
  name: z.string().min(1, 'Required').max(128),
  unit: z.string().min(1).max(32).optional(),
  reorder_level: z.number().int().min(0).optional(),
});
export type UpdateItemInput = z.infer<typeof updateItemSchema>;

/**
 * Free-form ledger movement. Receipts are excluded — they enter via
 * the gated receive flow (`/inventory/{id}/receive`), which pulls the
 * quantity from the item's `received` reorder request.
 *
 * `dispense` is now encounter-anchored (inventory audit fix): the
 * backend requires an open `encounter_id`, mirroring medicines.
 */
export const moveStockSchema = z
  .object({
    qty_delta: z.number().int().refine((v) => v !== 0, 'Delta must be non-zero'),
    reason_code: z.enum(['dispense', 'adjustment']),
    encounter_id: z.number().int().positive().optional(),
    note: z.string().max(255).optional(),
  })
  .refine(
    (v) => !(v.reason_code === 'dispense' && v.qty_delta > 0),
    { message: 'Delta sign does not match reason', path: ['qty_delta'] },
  )
  .refine(
    (v) => v.reason_code !== 'dispense' || v.encounter_id !== undefined,
    { message: 'Select the open encounter', path: ['encounter_id'] },
  );
export type MoveStockInput = z.infer<typeof moveStockSchema>;

/**
 * One row of the supply ledger (in/out with running balance). `qty_in`
 * / `qty_out` are mutually exclusive; `balance_after` is on-hand right
 * after the movement (panel revision — ledger-style tracking).
 * `reference_*` records which encounter/reorder the movement belonged
 * to; `user_email` is the actor (inventory audit fix).
 */
export const inventoryMovementSchema = z.object({
  id: z.number().int().positive(),
  reason_code: z.string(),
  qty_in: z.number().int().nullable(),
  qty_out: z.number().int().nullable(),
  balance_after: z.number().int().nullable(),
  reference_type: z.string().nullable(),
  reference_id: z.number().int().nullable(),
  user_email: z.string().nullable(),
  note: z.string().nullable(),
  created_at: z.string(),
});
export type InventoryMovement = z.infer<typeof inventoryMovementSchema>;

File: mobile/lib/core/models/inventory.dart
/// Clinic inventory (supplies) — mirrors `frontend/src/schemas/inventory.ts`
/// (`inventoryItemSchema`) and `InventoryItemDto`.
library;

class StockTransaction {
  StockTransaction({
    required this.id,
    required this.type,
    required this.createdAt,
    this.qtyIn,
    this.qtyOut,
    this.stockAfter,
    this.referenceType,
    this.referenceId,
    this.userEmail,
    this.note,
  });

  factory StockTransaction.fromJson(Map<String, dynamic> json) =>
      StockTransaction(
        id: (json['id'] ?? 0) as int,
        type: (json['type'] ?? json['reason_code'] ?? '') as String,
        qtyIn: (json['qty_in'] as num?)?.toInt(),
        qtyOut: (json['qty_out'] as num?)?.toInt(),
        stockAfter: (json['balance_after'] as num?)?.toInt(),
        referenceType: json['reference_type'] as String?,
        referenceId: json['reference_id'] as int?,
        userEmail: json['user_email'] as String?,
        note: json['note'] as String?,
        createdAt: (json['created_at'] ?? '') as String,
      );

  final int id;
  final String type;
  final int? qtyIn;
  final int? qtyOut;
  final int? stockAfter;
  final String? referenceType;
  final int? referenceId;
  final String? userEmail;
  final String? note;
  final String createdAt;
}

class InventoryMovementHint {
  InventoryMovementHint({
    required this.reasonCode,
    required this.qtyDelta,
    required this.createdAt,
    this.userEmail,
  });

  factory InventoryMovementHint.fromJson(Map<String, dynamic>? json) {
    if (json == null) {
      return InventoryMovementHint(
        reasonCode: '',
        qtyDelta: 0,
        createdAt: '',
      );
    }
    return InventoryMovementHint(
      reasonCode: (json['reason_code'] ?? '') as String,
      qtyDelta: (json['qty_delta'] ?? 0) as int,
      createdAt: (json['created_at'] ?? '') as String,
      userEmail: json['user_email'] as String?,
    );
  }

  final String reasonCode;
  final int qtyDelta;
  final String createdAt;
  final String? userEmail;
}

class InventoryItem {
  InventoryItem({
    required this.id,
    required this.sku,
    required this.name,
    required this.unit,
    required this.quantityOnHand,
    required this.reorderLevel,
    this.targetStock,
    required this.stockStatus,
    required this.lowStock,
    required this.archived,
    required this.createdAt,
    this.lastMovement,
  });

  factory InventoryItem.fromJson(Map<String, dynamic> json) => InventoryItem(
        id: json['id'] as int,
        sku: (json['sku'] ?? '') as String,
        name: (json['name'] ?? '') as String,
        unit: (json['unit'] ?? 'pc') as String,
        quantityOnHand: (json['quantity_on_hand'] ?? 0) as int,
        reorderLevel: (json['reorder_level'] ?? 0) as int,
        targetStock: json['target_stock'] as int?,
        stockStatus: (json['stock_status'] ??
            (((json['quantity_on_hand'] ?? 0) as int) <= 0
                ? 'out_of_stock'
                : (((json['quantity_on_hand'] ?? 0) as int) <=
                        ((json['reorder_level'] ?? 0) as int)
                    ? 'needs_to_reorder'
                    : 'in_stock'))) as String,
        lowStock: (json['low_stock'] ?? false) as bool,
        archived: (json['archived'] ?? false) as bool,
        createdAt: (json['created_at'] ?? '') as String,
        lastMovement: InventoryMovementHint.fromJson(
          json['last_movement'] as Map<String, dynamic>?,
        ),
      );

  final int id;
  final String sku;
  final String name;
  final String unit;
  final int quantityOnHand;
  final int reorderLevel;
  final int? targetStock;
  final String stockStatus;
  final bool lowStock;
  final bool archived;
  final String createdAt;

  /// Row-level last-move hint (null when there are no movements yet).
  final InventoryMovementHint? lastMovement;

  String get stockStatusLabel => switch (stockStatus) {
        'out_of_stock' => 'Out of Stock',
        'needs_to_reorder' => 'Needs to Reorder',
        _ => 'In Stock',
      };
}
